### DISCUSSION TITLE
How to I contribute?

### DISCUSSION THREAD

**Discussion by iliakan** [emscripten-core/emscripten#18734] [Q&A] [answered]
I'd like to contribute some code to emscripten, including tests.

Usually, to contribute to a project, I do it like this:
1) Github fork, clone.
2) Make changes, test locally, run install/build commands.
3) Git push && create PR.

Here the process looks to be different, can you describe it in a bit more details? 

In particular, I have troubles running the tests. Here's what I get by running directly from a cloned repo:
```bash
➜  emscripten git:(proxyfs-mmap) test/runner test_loop
posixtestsuite not found (run git submodule update --init?)
Test suites:
['test_core']
Running test_core: (1 tests)
(checking sanity from test runner)
runner: warning: LLVM version for clang executable "/usr/bin/clang" appears incorrect (seeing "14.0", expected "17") [-Wversion-check]
shared:CRITICAL: LLVM has not been built with the WebAssembly backend, clang reports:
...
```

Do I need a setup before running tests?

@sbc100 recommended to see https://emscripten.org/docs/contributing/developers_guide.html#developers-guide-setting-up. 

It's great to have such article, but it's quite confusing. 
- The article starts with a suggestion to run `emsdk`, but `emsdk` isn't a system utility.  Should I first install it? Looks like some steps are missing. 
- Then it suggests to install `tot-upstream`, but to contribute I'd really prefer a git clone. Is it possible? 

These things may be obvious for you, but not for a newcomer ;(

Please advise a step-by-step way to get emscripten from github, run tests and push the changes.

**Comment by sbc100** [comment, marked as answer]
Emscripten depends on both llvm and on binaryen.   The simplest way to get binaries for those tools is it use emsdk.

You can configure your local emscripten git checkout to use those either by setting up your own `.emscripten` config file to point to the emsdk version of these tools, or simply by using the emsdk config file directly with `export EM_CONFIG=/path/to/emsdk/.emscripten`.

If you want to build llvm and binaryen from source that is also an option but can take a long time and a lof of disk space, and is also unnecessary if you only want to contribute to emscirpten.

**Reply by iliakan** [reply]
I'm sorry, that doesn't answer my questions ;(
Maybe I asked them not clearly enough...

First, I need emsdk for the start, where to get it? =)

P.S. Thank you very much for your efforts. I really wonder, is there anyone else here?

**Reply by sbc100** [reply]
Instructions for installing using emsdk are here: https://emscripten.org/docs/getting_started/downloads.html#installation-instructions-using-the-emsdk-recommended

**Reply by iliakan** [reply]
Cool! So I go:
```bash
cd /opt/projects
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
./emsdk install latest
./emsdk activate latest
source ./emsdk_env.sh
```

EMSDK installed. 
The next step is suggested to be
```bash
emsdk install tot-upstream
emsdk activate tot-upstream
```

I tried, but it downloads zipped release.

Am I suggested to modify the downloaded code?

Can I use a similar command, but with a cloned emscripten repo (then one I expect to modify)?

**Reply by sbc100** [reply]
Yes, you can.   To use your cloned emscripten repo simply set `EM_CONFIG=/path/to/emsdk/.emscripten`, then run `emcc` from your cloned repo, either by putting your cloned repo first in your PATH to running emcc locally with `./emcc` or `./test/runner` (which always uses the local emcc).

**Reply by sbc100** [reply]
(To be clear you need to export the `EM_CONFIG` environment variable so it would be `export EM_CONFIG=/path/to/emsdk/.emscripten` on linux or mac).

**Reply by iliakan** [reply]
Ok, so here are my next steps:

```
cd /opt/projects
git clone https://github.com/emscripten-core/emscripten.git
export EM_CONFIG=/opt/projects/emsdk/.emscripten
cd emscripten
test/runner test_loop
```

PASSED! 🥳 

But then I try to run `test_proxyfs` from `test/test_other.py` - how to do that?

I tried `test/runner test_proxyfs` and `test/runner test_other` - both show errors:
```
posixtestsuite not found (run git submodule update --init?)
ERROR: could not find the following tests: core0.test_proxyfs
ERROR: could not find the following tests: core0.test_other.
```

**Reply by sbc100** [reply]
To the `other` tests you prefix the test name with `other`.  e.g. `./test/runner other.test_proxyfs`.   Or to run them all your would run `./test/runner other`.

**Reply by iliakan** [reply]
```
./test/runner other.test_proxyfs
```

Worked! 😮 🍾 🙇

**Reply by iliakan** [reply]
Thanks, it looks like I'm on the road to go further contributing 🎊
